Add arbitrary URL parameters to CUSTOM interfaces

solth opened the pull request:
This pull request adds the option to add arbitrary URL parameters to `ImportConfigurations` with search interface type `CUSTOM`. It also limits the number of search fields in `CUSTOM` search interfaces to one, because the current implemention for `CUSTOM` interfaces does not support hitlists, but only queries by ID (which should be sufficient in most cases anyway.) Fixes #5257 

@BartChris could you check whether or not this re-enables you to import from the simulated OAI interface you mentioned in #5271 ? If so, I will add a test to this pull request and mark it ready for review.

BartChris commented:
@solth Thanks a lot for your efforts. This works for my simulated OAI interface.

solth commented:
@BartChris would you mind adding a review for this pull request?

solth commented:
Unfortunately I don't know what's wrong with the Selenium test. Locally it always runs successful, but on GitHub it fails every time.

BartChris commented:
@solth yes, i will check it

BartChris commented on Kitodo/src/main/java/org/kitodo/production/forms/ImportConfigurationEditView.java:
```diff
@@ -225,6 +226,43 @@ public void removeSearchField(SearchField searchField) {
+    public void addUrlParameter(UrlParameter urlParameter) {
+        if (importConfiguration.getUrlParameters().stream().map(UrlParameter::getParameterKey)
+                .collect(Collectors.toList()).contains(urlParameter.getParameterKey())) {
+            Helper.setErrorMessage("URL parameter keys must be unique ('" + urlParameter.getParameterKey()
```
The error message is shown, but only showing for the fraction of a second.

solth replied:
This seems to have been caused by the `saveButtonToggler`. This hidden button is used to activate the `Save` button and is called whenever the properties of an object being edited - in this case the `ImportConfiguration` - change. And since the `error-messages` component was set to `autoUpdate`, the previously displayed "URL parameter keys must be unique" message would be immediately replaced with an empty message array when the `saveButtonToggler` was triggered. 

To resolve the issue, I replaced the `error-messages` component with `autoUpdate` with an identical `error-messages` component _without_ the `autoUpdate` in the `importConfigurationEdit` view.

BartChris commented on Kitodo/src/test/java/org/kitodo/production/services/data/ImportServiceIT.java:
```diff
@@ -83,31 +91,73 @@ public static void cleanDatabase() throws Exception {
 
-    private static void setupServer(String serverResponse) {
+    @Test
+    public void shouldCreateUrlWithCustomParameters() throws DAOException, ImportException, IOException {
```
The naming does not fully describe what is happeing in the test as the creation of url parameter is happening in the MockDatabase class and not in this test. Maybe this should be renamed to reflect what is happening in the test (`shouldCreateProcessWithCustomUrlParameters.`)  and also check if the imported processes is the expected one and not just "not null"

solth replied:
Actually, the MockDatabase only prepares the ImportConfiguration with custom URL parameters. This test does not check the creation of the _parameters_, though, but instead the creation of the _URL_ itself, using the parameters of the ImportConfiguration. For that reason I think the method is named correctly. 

The creation of an ImportConfiguration with custom URL parameters on the other hand is tested in the Selenium Test `AddingST.addCustomImportconfigurationWithUrlParameters`.

BartChris commented:
@solth Functionality works good for me. I added some comments.

BartChris commented:
> Unfortunately I don't know what's wrong with the Selenium test. Locally it always runs successful, but on GitHub it fails every time.

The selenium test is also failing locally for me

<pre>
[ERROR] addCustomImportconfigurationWithUrlParameters  Time elapsed: 17.862 s  <<< FAILURE!
org.opentest4j.MultipleFailuresError: 
Multiple Failures (2 failures)
        org.awaitility.core.ConditionTimeoutException: Condition with lambda expression in org.kitodo.selenium.testframework.pages.ImportConfigurationEditPage was not fulfilled within 5 seconds.
        org.openqa.selenium.UnhandledAlertException: unexpected alert open: {Alert text : }
  (Session info: chrome=106.0.5249.119): {Alert text : 
</pre>

solth commented on Kitodo-DataManagement/src/main/java/org/kitodo/data/database/persistence/UrlParameterDAO.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    @Override
+    public List<UrlParameter> getAll(int offset, int size) throws DAOException {
+        return retrieveObjects("FROM urlparameter ORDER BY ID ASC", offset, size);
```
The name of the database table is `urlparameter` (all lowercase characters), not `UrlParameter` (camal case). Why do you suggest to change the parameter in the query to camel case?

BartChris replied:
I have not tested it, but i think in HQL you have to use the name of the Entity, not the table name. This is also done for the getAll-method in other DAO-classes.

solth replied:
You are right! I just tested the method and it did indeed throw the following exception:
```org.hibernate.hql.internal.ast.QuerySyntaxException: urlparameter is not mapped [FROM urlparameter ORDER BY ID ASC]```
Implementing your change request does resolve the problem!

solth commented:
> Edit: The test seems to use the wrong ID for the button. I changed
> 
> https://github.com/kitodo/kitodo-production/blob/128bdb2097d0a65cf275db739f89877238787a45/Kitodo/src/test/java/org/kitodo/selenium/testframework/pages/ImportConfigurationEditPage.java#L48-L49

Great! Thanks alot for looking into the failing test!

solth commented on Kitodo/src/main/webapp/WEB-INF/templates/includes/importConfigurationEdit/rows/searchFieldRows.xhtml:
```diff
@@ -72,8 +72,11 @@
                            value="#{msgs['tooltip.importConfig.searchFieldHelp']}"/>
                 <p:commandButton icon="fa fa-plus"
+                                 id="addSearchField"
+                                 title="#{(importConfigurationEditView.importConfiguration.interfaceType eq 'CUSTOM' and importConfigurationEditView.importConfiguration.searchFields.size() > 0) ? msgs['importConfig.custom.idSearchOnly'] : msgs['importConfig.addSearchField']}"
```
Yes, we had the same issue with some other deactivated command components in the past. Fixed it by wrapping the button in a separate container element and moving the tooltip from the button to the wrapper. Thanks for pointing this out.

solth commented:
> ```java
> ```java
> 
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> After that the tests finish succesful. Is using a static ID here the only possibility and will this break again when other changes are introduced to the form?
> ```

I think the best way is to explicitely set the ID to a fixed value and use that ID in the Selenium test because then the ID won't change with new form elements being added later on.

solth commented:
@BartChris thanks a lot for the review. I implemented your change requests.

solth commented:
@BartChris would you mind reviewing this PR?

BartChris commented on Kitodo/src/main/java/org/kitodo/production/forms/ImportConfigurationEditView.java:
```diff
@@ -225,6 +226,43 @@ public void removeSearchField(SearchField searchField) {
+    public void addUrlParameter(UrlParameter urlParameter) {
+        if (importConfiguration.getUrlParameters().stream().map(UrlParameter::getParameterKey)
+                .collect(Collectors.toList()).contains(urlParameter.getParameterKey())) {
+            Helper.setErrorMessage("URL parameter keys must be unique (\"" + urlParameter.getParameterKey()
```
error message should probably be translated.

BartChris replied:
is it possible to translate this error message as well?

BartChris commented:
I think i discovered another bug: When i try to change a url parameter, all of the defined parameters are changed and not just the one i want to change.

BartChris reviewed (changes requested):
Updating url params is broken

solth commented:
> I think i discovered another bug: When i try to change an url parameter, all of the defined parameters (if i have multiple defined) are changed and not just the one i want to change.

Thanks a lot for discovering this nasty bug. The same seems to be true for _unsaved_ search fields.

solth commented:
@BartChris I added message keys for the error message you mentioned and fixed the bug when updating URL parameters. Could you test it once more? The similar bug with the search fields seems to be caused by something different and isn't fixed yet, but I would suggest to deal with that in a separate pull request because it isn't really connected to the topic of this one.

BartChris reviewed (approved):
Looks good to me!

solth commented:
Thanks for the review. I added a new issue for bug when updating unsaved search fields: #5420
